### ISSUE TITLE
systemd not always producing the %t from systemd file

### ISSUE THREAD

**Issue by ikke-t** [containers/podman#5572]
/kind bug

**Description**

I followed the blog to use conmon-pidfile and cidfile from [this blog](https://www.redhat.com/sysadmin/podman-shareable-systemd-services). However, systemd doesn't fills %t for podman user containers.

ping @vrothberg 

**Steps to reproduce the issue:**

1. Ansible does this systemd file:

```
cat /etc/systemd/system/ha-container-pod.service 
[Unit]
Description=ha Podman Container
After=network.target

[Service]
Type=simple
TimeoutStartSec=15
ExecStartPre=-/usr/bin/rm -f /%t/%n-pid /%t/%n-cid
User=redhatter

ExecStart=/usr/bin/podman run --name ha \
  --rm -p 8123:8123/tcp -v "/var/lib/containers/storage/homeassistant-config:/config:Z" --hostname="ha.localnet" --memory=1024M \
  --conmon-pidfile  /%t/%n-pid --cidfile /%t/%n-cid \
  homeassistant/home-assistant:stable

ExecStop=/usr/bin/sh -c "/usr/bin/podman stop -t "15" `cat /%t/%n-cid`"
ExecStop=/usr/bin/sh -c "/usr/bin/podman rm -f `cat /%t/%n-cid`"
Restart=on-failure
RestartSec=30
KillMode=none
PIDFile=/%t/%n-pid

[Install]
WantedBy=multi-user.target

```

2 add user ```sudo useradd redhatter```
3. reload systemd ```sudo systemctl daemon-reload```
4. start container:  sudo systemctl start ha-container-pod.service

**Describe the results you received:**

```
sudo journalctl -eu ha-container-pod.service
Mar 20 17:12:57 gr8.localdomain systemd[1]: Started ha Podman Container.
Mar 20 17:12:57 gr8.localdomain podman[11501]: Error: error opening cidfile //run/ha-container-pod.service-cid
Mar 20 17:12:57 gr8.localdomain systemd[1]: ha-container-pod.service: Main process exited, code=exited, status=125/n/a
Mar 20 17:12:57 gr8.localdomain systemd[1]: ha-container-pod.service: Failed with result 'exit-code'.
Mar 20 17:13:27 gr8.localdomain systemd[1]: ha-container-pod.service: Service RestartSec=30s expired, scheduling restart.
Mar 20 17:13:27 gr8.localdomain systemd[1]: ha-container-pod.service: Scheduled restart job, restart counter is at 46.
Mar 20 17:13:27 gr8.localdomain systemd[1]: Stopped ha Podman Container.
```

**Describe the results you expected:**

start the container with proper path in cidfile

**Additional information you deem important (e.g. issue happens only occasionally):**

always

**Output of `podman version`:**

```
Version:            1.6.4
RemoteAPI Version:  1
Go Version:         go1.12.12
OS/Arch:            linux/amd64
```

**Output of `podman info --debug`:**

```
debug:
  compiler: gc
  git commit: ""
  go version: go1.12.12
  podman version: 1.6.4
host:
  BuildahVersion: 1.12.0-dev
  CgroupVersion: v1
  Conmon:
    package: conmon-2.0.6-1.module+el8.1.1+5259+bcdd613a.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.0.6, commit: 6ffbb2ec70dbe5ba56e4bfde946fb04f19dd8bbf'
  Distribution:
    distribution: '"rhel"'
    version: "8.1"
  MemFree: 526786560
  MemTotal: 16276774912
  OCIRuntime:
    name: runc
    package: runc-1.0.0-64.rc9.module+el8.1.1+5259+bcdd613a.x86_64
    path: /usr/bin/runc
    version: 'runc version spec: 1.0.1-dev'
  SwapFree: 8167911424
  SwapTotal: 8292134912
  arch: amd64
  cpus: 8
  eventlogger: journald
  hostname: gr8.localdomain
  kernel: 4.18.0-147.5.1.el8_1.x86_64
  os: linux
  rootless: false
  uptime: 76h 28m 48.87s (Approximately 3.17 days)
registries:
  blocked: null
  insecure: null
  search:
  - registry.access.redhat.com
  - registry.redhat.io
  - quay.io
  - docker.io
store:
  ConfigFile: /etc/containers/storage.conf
  ContainerStore:
    number: 3
  GraphDriverName: overlay
  GraphOptions: {}
  GraphRoot: /var/lib/containers/storage
  GraphStatus:
    Backing Filesystem: xfs
    Native Overlay Diff: "true"
    Supports d_type: "true"
    Using metacopy: "false"
  ImageStore:
    number: 18
  RunRoot: /var/run/containers/storage
  VolumePath: /var/lib/containers/storage/volumes
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
podman-1.6.4-2.module+el8.1.1+5363+bf8ff1af.x86_64
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
rhel8.1 laptop

**Comment by rhatdan** [comment]
@vrothberg PTAL

**Comment by ikke-t** [comment]
It's related to this ansible role. I changed %t to %T to make it work somehow. https://github.com/ikke-t/podman-container-systemd

**Comment by vrothberg** [comment]
Thanks for opening the issue and apologies for my silence! I am currently focusing on Podman v2 but will get back to systemd stuff in May.

**Comment by cajus** [comment]
Same with the latest version. If I replace %t with i.e. _/run/users/1234_ podman is segfaulting while it works fine if started as the same user by `systemctl --user`.

**Comment by storrgie** [comment]
I'd also suggest, even if it seems weird, that there is a `ExecStopPost=/usr/bin/rm -f %T/%N.cid`. It seems as if enumerating to systemd via `PIDFile` automatically cleans up that `.pid`, but the `.cid` is left around. I see that by default the template attempts to clear it upon `ExecStartPre`, but it feels somewhat disingenuous that systemd would clear the `.pid` at the end of a unit run and we'd leave that `.cid` around for the next restart of the unit.

More discussion on this in #6157.

**Comment by vrothberg** [comment]
To clarify the `.pid` removal. It's not systemd removing but Podman+conmon, so the inconsistency is on our side.

While it should not cause any harm, I find the inconsistency at least confusing. Unfortunately, I don't think we can change the behavior without (potentially) breaking users.

**Comment by vrothberg** [comment]
Note to self: make sure that the file removal is properly explained in the man pages.

**Comment by storrgie** [comment]
> To clarify the `.pid` removal. It's not systemd removing but Podman+conmon, so the inconsistency is on our side.
> 
> While it should not cause any harm, I find the inconsistency at least confusing. Unfortunately, I don't think we can change the behavior without (potentially) breaking users.

I'd venture that one of the __big strengths__ of podman is rootless/daemon-less with combination to your existing init system. So any amount of consistency that can be achieved would be welcome as quality of life when combining podman+systemd . Documenting, while helpful, is going to only help comprehensive RTFM momements, when really this idiosyncrasy is likely to catch and confuse people for some time before they discover it documented as an idiosyncrasy in the doc.

**Comment by vrothberg** [comment]
Thanks for the feedback! I cannot disagree but I need to keep backwards compat in the back of my head.

@mheon @baude @rhatdan , what are your thoughts on the issue? In summary, the --conmon-pidfile is being removed with the container while the --cidfile is not.

**Comment by mheon** [comment]
`--cidfile` is a deliberate Docker compatability measure, AFAIK - we made the lifecycle match what Docker expected. They're also entirely outside of Libpod - we don't know that they exist, so once they're made we never remove them. We had no such constraints with pidfiles - they're fully managed by Libpod, and required to interact with init systems (and, since a while ago, used for every container). I think we also need to remove them to ensure that Conmon starts cleanly (it blows up when they still exist).

**Comment by rhatdan** [comment]
Can you remove them in the generated unit file?

**Comment by vrothberg** [comment]
> Can you remove them in the generated unit file?

Yes, we can do that in the unit file as well.

**Comment by storrgie** [comment]
> > Can you remove them in the generated unit file?
> 
> Yes, we can do that in the unit file as well.

Right, the concern I'd had was that `.pid` was being removed implicitly (by conman, I'd thought systemd) but the unit file was also attempting to remove it in `ExecStartPre` (assuming was left there as cruft or in case of conman crash?).

I'd ended up here in my futzing around:
```
[Unit]
Description=Podman container-mattermost-postgres.service
Wants=network.target
After=network-online.target

[Service]
WorkingDirectory=/opt/mattermost
User=mattermost
Group=mattermost
Environment=PODMAN_SYSTEMD_UNIT=%N
Restart=on-failure
ExecStartPre=/usr/bin/rm -f %T/%N.pid %T/%N.cid
ExecStart=/usr/bin/podman run --conmon-pidfile %T/%N.pid --cidfile %T/%N.cid --cgroups=no-conmon \
  --name=postgres12 \
  --env-file /opt/mattermost/postgres.env \
  --net=host --publish 127.0.0.1:5432:5432 \
  --volume /opt/mattermost/postgres:/var/lib/postgresql/data:Z \
  docker.io/postgres:12
ExecStop=/usr/bin/podman stop --ignore --cidfile %T/%N.cid -t 10
ExecStopPost=/usr/bin/podman rm --ignore -f --cidfile %T/%N.cid
ExecStopPost=/usr/bin/rm -f %T/%N.cid
PIDFile=%T/%N.pid
KillMode=none
Type=simple

[Install]
WantedBy=multi-user.target default.target
```
 - I'm specifically __not__ detaching because I want console output to be caught by systemd/journald, which is accomplished by removing the `-d` as well as changing `Type` to simple.
 - I'm using `%T` instead of `%t` which makes this more transportable to be initialized by root
 - I'm using the `User`, `Group`, and `WorkingDirectory` to run rootless as a user
 - I'm removing the `.cid` at both unit stop with an extra `ExectStopPost` (which seems to work consistently, but also might be bad form)
 - I've get the `.pid` and `.cid` removal in `ExecStartPre` as a sort of... scared case I guess.

**Comment by vrothberg** [comment]
We're working on this issue and are talking to the system developers. There are couple of issues with changing the user (i.e., using `User=`). Among other things, we can't use `PIDFile` as systemd will refuse it as it's being created by the non-root user.

We'll keep you posted.

**Comment by rhatdan** [comment]
@vrothberg is this still an issue?

**Comment by vrothberg** [comment]
Yes. The systemd team is investigating the issue.

**Comment by jdoss** [comment]
> Yes. The systemd team is investigating the issue.

Do they have an open issue that you know of that users can subscribe to?

**Comment by vrothberg** [comment]
@msekletar is investigating the issue. @msekletar, is there an issue regarding `PIDFile` issue (i.e., USER= and rootless podman creating the pidfile)?

**Comment by goochjj** [comment]
FWIW, I believe the best solution to this is to use #6693  and --sdnotify conmon.  It makes passing the PIDFile to systemd unnecessary, works with User=, root and rootless, works around issues in RHEL with "the pid isn't owned by root".

Your units then change to:
- Add --sdnotify conmon
- Remove all .pid and .cid overrides.  Conmon already writes them to the XDG_RUNTIME container dir anyway
- Remove PIDFile=
- Change Type to notify
- Add NotifyAccess=all (I haven't tested if this is necessary or could be more restrictive)

**Comment by goochjj** [comment]
And it makes the unit look a lot cleaner. You can get rid of all the explicit rms and cidfile reads as long as you --name %N (or something you know)

**Comment by ikke-t** [comment]
Looks much clearer solution, thanks. I need to wait that it get's into podman and into RHEL.

**Comment by vrothberg** [comment]
> And it makes the unit look a lot cleaner. You can get rid of all the explicit rms and cidfile reads as long as you --name %N (or something you know)

I concur. Thanks for your great work, this really makes things _much_ simpler.

**Comment by rhatdan** [comment]
@ikke-t @vrothberg Is this merged?  Can we close this issue?

**Comment by vrothberg** [comment]
It is still an issue. We should migrate the systemd unit files over to sdnotify. I hope to find time once we've burned down the number of issues and the short-name work is done.

**Comment by rhatdan** [comment]
@vrothberg Any time to work on this one?

**Comment by vrothberg** [comment]
> @vrothberg Any time to work on this one?

I think we need to wait until the work in conmon is done.

**Comment by rhatdan** [comment]
@haircommander @vrothberg Where are we on this one?

**Comment by vrothberg** [comment]
No progress yet

**Comment by holgerpieta** [comment]
Not sure if that's helpful, but I tried to get it going on fully updated Raspberry PI OS 10 using the notify option:
```
[Unit]
Description=Podman container-syncthing.service
Documentation=man:podman-generate-systemd(1)
Wants=network.target
BindsTo=mnt-data.mount
After=network-online.target mnt-data.mount

[Service]
Type=notify
NotifyAccess=all
User=pi
Environment=PODMAN_SYSTEMD_UNIT=%n
ExecStart=/usr/bin/podman run --sdnotify=conmon --cgroups=no-conmon -d --replace --arch=arm --variant=v7 --name=syncthing --label io.containers.autoupdate=imag\
e -p 8384:8384 -p 22000:22000/tcp -p 22000:22000/udp -p 21027:21027/udp -v /mnt/data/:/var/syncthing -e TZ=Europe/Berlin docker.io/syncthing/syncthing:latest

[Install]
WantedBy=multi-user.target default.target
```
(I know important bits are missing, e.g. ExecStop, but that's a problem for later.)
It kind of works, as in starting the container, but systemd is not happy:
`Job for container-syncthing-notify.service failed because the service did not take the steps required by its unit configuration.`
I think this is the important bit in the journal:
`user@1000.service: Got notification message from PID 1643, but reception only permitted for main PID 507`
PID 1643 is the conmon, so that's corrent.
But I think the problem is not the PID, but the control group. As far as I can see, the conmon process is in the user slice control group while systemd is in the system slice. And thus systemd doesn't like to receive the notify. The man page quite explicitly states that systemd only receives notification from process in the same control group.

Anyone knows enough about systemd and control groups to give me hints for debugging that further.
I have a bit of a hunch that it is never going to work, because a user process should not be able to switch back into the system slice, right?

**Comment by vrothberg** [comment]
I finally found some time to tackle the issue and opened https://github.com/containers/podman/pull/10557.

**Comment by mmaguigan** [comment]
@vrothberg referring to the comment by @holgerpieta I am also having the same problem with the proposed solution on `CentOS 8.4.2105`. This causes systemd to mark the job as failed with the result `protocol`...

**Comment by vrothberg** [comment]
@mmaguigan, thanks! Would you open a new issue for that?

**Comment by developedsoftware** [comment]
@holgerpieta did you ever resolve this? 

I am having the exact same issue. Is there a solution? 

`user@1001.service: Got notification message from PID 632662, but reception only permitted for main PID 1530`

**Comment by holgerpieta** [comment]
In a way, yes. Maybe not the nicest solution and somewhat bypassing the way it's supposed to work, but it works:
```
[Unit]
Description=Podman container-syncthing.service
Documentation=man:podman-generate-systemd(1)
Wants=network.target
BindsTo=systemd-cryptsetup@cryptdata.service mnt-data.mount
After=network-online.target systemd-cryptsetup@cryptdata.service mnt-data.mount

[Service]
Type=exec
User=syncthing
NoNewPrivileges=true
WorkingDirectory=~
PIDFile=%T/%N.pid
Environment=PODMAN_SYSTEMD_UNIT=%n
ExecStartPre=/usr/bin/rm -f %T/%N.pid %T/%N.cid
ExecStart=/usr/bin/podman run --conmon-pidfile %T/%N.pid --cidfile %T/%N.cid --cgroups=no-conmon --replace --name=syncthing --label io.containers.autoupdate=image --network=slirp4netns:enable_ipv6=true -p 8384:8384 -p 22000:22000/tcp -p 22000:22000/udp -p 21027:21027/udp -v /mnt/data/:/var/syncthing -e TZ=Europe/Berlin docker.io/syncthing/syncthing:latest
ExecStop=/usr/bin/podman stop --ignore --cidfile %T/%N.cid -t 10
ExecStopPost=/usr/bin/podman rm --ignore -f --cidfile %T/%N.cid
ExecStopPost=/usr/bin/rm -f %T/%N.cid
KillMode=none
#Restart=on-failure
TimeoutStopSec=70
````

Recently I have the problem that I cannot start the container unless after I `su`'ed to `syncthing` and called `podman unshare cat /proc/self/uid_map`. Not sure why that is.
